stdpopsim/models.py

```diff
@@ -137,6 +137,153 @@ def asdict(self):
+        time axis are returned.
+
+        Returns a tuple where the first value
+        is the steps (in generations) and the second is a numpy
+        ndarray that will contain the coalescence rate for
+        the entire population.
+        """
+
+        num_pops = self.num_pops()
+        assert(len(num_samples) == num_pops)
+        P = np.zeros([num_pops**2, num_pops**2])
+        index_array = np.array(range(num_pops**2)).reshape([num_pops, num_pops])
+        for x in range(num_pops):
+            for y in range(num_pops):
+                K_x = num_samples[x]
+                K_y = num_samples[y]
+                P[index_array[x, y], index_array[x, y]] = K_x * (K_y - (x == y))
+        P = P / np.sum(P)
+        r = np.zeros(num_steps)
+        steps = np.linspace(0, end, num_steps)
+        dt = steps[1] - steps[0]
+        mass_migration_objects = []
+        mass_migration_steps = []
+        for demo in self.demographic_events:
+            if type(demo) == msprime.simulations.MassMigration:
+                mass_migration_objects.append(demo)
+                mass_migration_steps.append(int(demo.time/dt))
+        for j in range(num_steps):
+            N_M = self.pop_size_and_migration_at_t(steps[j])
+            N = N_M["population_sizes"]
+            M = N_M["migration_matrix"]
+            C = np.zeros([num_pops**2, num_pops**2])
+            index_array = np.array(range(num_pops**2)).reshape([num_pops, num_pops])
+            for idx in range(num_pops):
+                C[index_array[idx, idx], index_array[idx, idx]] = 1 / (2 * N[idx])
+            for idx, row in enumerate(M):
+                M[idx][idx] = -1 * sum(row)
+            Identity = np.eye(len(M[0]))
+            G = (np.kron(M, Identity) + np.kron(Identity, M)) - C
+            if j in mass_migration_steps:
+                idx = mass_migration_steps.index(j)
+                ds = mass_migration_objects[idx].time - (j * dt)
+                a = mass_migration_objects[idx].source
+                b = mass_migration_objects[idx].dest
+                p = mass_migration_objects[idx].proportion
+                J = np.array(range(num_pops**2)).reshape([num_pops, num_pops])
+                S = np.eye(num_pops**2, num_pops**2)
+                for x in range(num_pops):
+                    if x == a:
+                        S[J[a, a], J[a, b]] = S[J[a, a], J[b, a]] = p * (1 - p)
+                        S[J[a, a], J[b, b]] = p ** 2
+                    else:
+                        S[J[x, a], J[x, b]] = S[J[a, x], J[b, x]] = p
+                        S[J[x, a], J[x, a]] -= p
+                        S[J[a, x], J[a, x]] -= p
+                P = np.matmul(P, S)
+                P = np.matmul(P, scipy.linalg.expm((dt - ds) * G))
+            else:
+                P = np.matmul(P, scipy.linalg.expm(dt * G))
+            r[j] = np.sum(np.matmul(P, C)) / np.sum(P)
+        return steps, 1/(2*r)
```

Comment:
This function is called `coalescence_rate_trajctory` but it returns "effective population sizes" instead of coalescence rates. I vote to have this function return `steps, r` and let the user plot `1/(2*r)` if that's what they want.